core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,325 @@
+}
+
+static int
+toStandardGamepadButton(uint16_t manetteButton)
+{
+    switch (manetteButton) {
+    case BTN_SOUTH:
+        return WPE_GAMEPAD_BUTTON_BOTTOM;
+    case BTN_EAST:
+        return WPE_GAMEPAD_BUTTON_RIGHT;
+    case BTN_NORTH:
+        return WPE_GAMEPAD_BUTTON_TOP;
+    case BTN_WEST:
+        return WPE_GAMEPAD_BUTTON_LEFT;
+    case BTN_TL:
+        return WPE_GAMEPAD_BUTTON_LEFT_SHOULDER;
+    case BTN_TR:
+        return WPE_GAMEPAD_BUTTON_RIGHT_SHOULDER;
+    case BTN_TL2:
+        return WPE_GAMEPAD_BUTTON_LEFT_TRIGGER;
+    case BTN_TR2:
+        return WPE_GAMEPAD_BUTTON_RIGHT_TRIGGER;
+    case BTN_SELECT:
+        return WPE_GAMEPAD_BUTTON_SELECT;
+    case BTN_START:
+        return WPE_GAMEPAD_BUTTON_START;
+    case BTN_THUMBL:
+        return WPE_GAMEPAD_BUTTON_LEFT_STICK;
+    case BTN_THUMBR:
+        return WPE_GAMEPAD_BUTTON_RIGHT_STICK;
+    case BTN_DPAD_UP:
+        return WPE_GAMEPAD_BUTTON_D_PAD_TOP;
+    case BTN_DPAD_DOWN:
+        return WPE_GAMEPAD_BUTTON_D_PAD_BOTTOM;
+    case BTN_DPAD_LEFT:
+        return WPE_GAMEPAD_BUTTON_D_PAD_LEFT;
+    case BTN_DPAD_RIGHT:
+        return WPE_GAMEPAD_BUTTON_D_PAD_RIGHT;
+    case BTN_MODE:
+        return WPE_GAMEPAD_BUTTON_CENTER;
+    default:
+        return -1;
+    }
+}
+
+static void
+on_button_press(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    uint16_t button;
+    int      std_button;
+
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    std_button = toStandardGamepadButton(button);
+    if (std_button != -1)
+        wpe_gamepad_dispatch_button_changed(gamepad->server, (enum wpe_gamepad_button) std_button, true);
+}
+
+static void
+on_button_release(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    uint16_t button;
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    wpe_gamepad_dispatch_button_changed(gamepad->server, toStandardGamepadButton(button), false);
+}
+
+static void
+device_disconnect_cb(ManetteDevice *device, Gamepad *gamepad)
+{
+    if (!gamepad)
+        return;
+
+    if (gamepad->button_pressed > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_pressed);
+        gamepad->button_pressed = 0;
+    }
+
+    if (gamepad->button_released > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_released);
+        gamepad->button_released = 0;
+    }
+
+    if (gamepad->axis_event > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->axis_event);
+        gamepad->axis_event = 0;
+    }
+
+    if (gamepad->disconnected > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->disconnected);
+        gamepad->disconnected = 0;
+    }
+}
+
+static gpointer
+gamepad_create(struct wpe_gamepad *server, unsigned gamepad_id)
+{
+    Gamepad       *gamepad;
+    ManetteDevice *device;
+
+    device = GUINT_TO_POINTER(gamepad_id);
+    if (!MANETTE_IS_DEVICE(device))
```

Comment:
This should never happen, because we control how identifiers are generated and they are always valid pointers. If you feel like adding some safeguard here, use `g_assert(MANETTE_IS_DEVICE(device))` instead, right?

Reply:
done as you suggested.